File: backend/module-maintenance/src/main/java/univ/lille/module_maintenance/domain/model/Ticket.java
package univ.lille.module_maintenance.domain.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Builder
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Ticket {
    private Long id;
    
    private String title;
    private String description;
    private Priority priority;
    
    private Status status;
        
    private Long createdByUserId;
    private String createdByUserName;
    private Long organizationId;
    
    @Builder.Default
    private List<Comment> comments = new ArrayList<>();
    
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    
    public boolean belongsTo(Long userId) {
        return this.createdByUserId != null && this.createdByUserId.equals(userId);
    }

    public boolean belongsToOrganization(Long orgId) {
        return this.organizationId != null && this.organizationId.equals(orgId);
    }

    public void updateStatus(Status newStatus) {
        if (this.status.canTransitionTo(newStatus)) {
            this.status = newStatus;
            this.updatedAt = LocalDateTime.now();
        } else {
            throw new IllegalStateException("Cannot transition from " + this.status + " to " + newStatus);
        }
    }
    
    public void addComment(Comment comment) {
        this.comments.add(comment);
        this.updatedAt = LocalDateTime.now();
    }

    public boolean isOpen() {
        return this.status == Status.OPEN;
    }
    
    public boolean isInProgress() {
        return this.status == Status.IN_PROGRESS;
    }
    
    public boolean isResolved() {
        return this.status == Status.RESOLVED;
    }

    public boolean isRejected() {
        return this.status == Status.REJECTED;
    }
}

File: mobile/lib/features/maintenance/data/models/ticket.dart

import 'package:mobile/features/maintenance/data/models/status.dart';
import 'package:mobile/features/maintenance/data/models/priority.dart';
import 'package:mobile/features/maintenance/data/models/comment.dart';

class Ticket {
  final int? id;
  final String title;
  final String description;
  final Priority priority;
  final Status status;
  final int createdByUserId;
  final String createdByUserName;
  final int organizationId;
  final List<Comment> comments;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Ticket({
    this.id,
    required this.title,
    required this.description,
    required this.priority,
    required this.status,
    required this.createdByUserId,
    required this.createdByUserName,
    required this.organizationId,
    this.comments = const [],
    required this.createdAt,
    this.updatedAt,
  });

  factory Ticket.fromJson(Map<String, dynamic> json) {
    return Ticket(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      priority: Priority.values.byName((json['priority'] as String).toLowerCase()),
      status: Status.values.byName((json['status'] as String).toLowerCase()),
      createdByUserId: json['createdByUserId'],
      createdByUserName: json['createdByUserName'],
      organizationId: json['organizationId'],
      comments: (json['comments'] as List<dynamic>?)
              ?.map((c) => Comment.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt:
          json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'priority': priority.toString().split('.').last.toUpperCase(),
      'status': status.toString().split('.').last.toUpperCase(),
      'createdByUserId': createdByUserId,
      'createdByUserName': createdByUserName,
      'organizationId': organizationId,
      'comments': comments.map((c) => c.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  bool isOpen() => status == Status.open;
  bool isInProgress() => status == Status.inProgress;
  bool isResolved() => status == Status.resolved;
  bool isRejected() => status == Status.rejected;
  bool isCancelled() => status == Status.cancelled;

  bool belongsTo(int userId) => createdByUserId == userId;
  bool belongsToOrganization(int orgId) => organizationId == orgId;
}

